src/operator/distinct.rs

```diff
@@ -463,132 +521,205 @@ where
+                eprintln!("ts: {ts:?}, epoch_end: {:?}", self.time.epoch_end(scope));
+            }
+            !ts.less_equal(&self.time.epoch_end(scope))
+        }));
+
+        self.time = self.time.advance(scope + 1);
     }
 
     fn fixedpoint(&self, scope: Scope) -> bool {
-        // TODO: generalize `DistinctTrace` to support arbitrarily nested scopes.
-        assert_eq!(scope, 0);
+        let epoch_end = self.time.epoch_end(scope);
+        /*println!("time: {:?} epoch_end: {:?} fixedpoint: empty_output: {}, empty_input: {}, keys_of_interest: {}"
+        , self.time
+        , epoch_end
+        , self.empty_output
+        , self.empty_input
+        , self.keys_of_interest.keys()
+           .all(|ts| !ts.less_equal(&epoch_end)));*/
+
         self.empty_input
             && self.empty_output
-            && self.future_updates.values().all(|vals| vals.is_empty())
+            && self
+                .keys_of_interest
+                .keys()
+                .all(|ts| !ts.less_equal(&epoch_end))
     }
 }
 
-impl<Z, T> BinaryOperator<Z, T, Z> for DistinctTrace<Z, T>
+impl<Z, T> BinaryOperator<Z, T, Z> for DistinctIncremental<Z, T>
 where
-    Z: ZSet,
+    Z: IndexedZSet,
     Z::R: ZRingValue,
-    T: Trace<Key = Z::Key, Val = (), Time = NestedTimestamp32, R = Z::R>,
+    T: Trace<Key = Z::Key, Val = Z::Val, R = Z::R>,
 {
-    // Distinct does not have nice properties like linearity that help with
-    // incremental evaluation: computing an update for each key requires inspecting
-    // the entire history of this key.  Fortunately, we do not need to look at all
-    // keys in the trace.  In particular, only those keys that appeared in input
-    // deltas during the current clock epoch can appear in the output of this
-    // operator with non-zero weights.  So, as a simple heuristic, we could track
-    // the set of keys that appeared since the start of the epoch and only consider
-    // those.
-    //
-    // But we can restrict the set of keys to consider even more by using a stronger
-    // property: only the keys `k` that satisfy one of the following conditions can
-    // appear in the output of the operator at local time `t`:
-    // * keys in the current input `delta`
-    // * keys from earlier inputs observed during the current clock epoch that
-    //   appeared in one of the previous epochs at time `t`.
-    //
-    // To efficiently compute keys that satisfy the second condition, we use the
-    // `future_updates` map, where for each key observed in the current epoch
-    // at time `t1` we lookup the smallest time `t2 > t1` (if any) at which we saw
-    // the key during any previous epochs and record this key in
-    // `future_updates[t2]`. Then when evaluating an operatpr at time `t` we
-    // simply scan all keys in `delta` and all keys in `future_updates[t]` for
-    // candidates.
-    //
-    // TODO: Add example.
-    //
     fn eval(&mut self, delta: &Z, trace: &T) -> Z {
```

Comment:
It seems like an `eval_owned()` implementation could be beneficial for this operator since it clones inputs

Reply:
We'll never get an owned `trace` (it's always owned by the Z-1 operator), but we can get an owned `delta`, and yeah, that should allow us to reduce cloning. I'll add a todo.